Feature #1225 generate pdf

maartenhammer opened the pull request:
### Linked issue

Closes #1225

### How to test

You can run the pdf generator using [x, config] =  xASL_qc_GenerateReport(x[, subject]), with x being the ExploreASL s struct, if no subject is provided, it will default to the subject defined in x.SUBJECT.  
A [Small manual](https://github.com/ExploreASL/ExploreASL/blob/f5bb987cd9a04ddf6e7fe2a594a38af4cee17c0d/WorkInProgress/QualityControl/PDFGenerator.md) has been written that goes over the basics to creating a config report for the layout. 
If no configuration file is found in `<dataroot>/Derivatives/ExploreASL` it will use the default in the exploreASL root directory.

**Manual should be moved to the docs before merging!**

maartenhammer committed:
#1225 pre-release of new QC-PDF

maartenhammer committed:
#1225 changed to a line based approach

maartenhammer committed:
#1225 refactor config load

maartenhammer committed:
#1225 refactor Output load

maartenhammer committed:
#1225 added patient info and placeholder images

maartenhammer committed:
#1225 cleanup

maartenhammer committed:
#1225 Added functions (wip) for slab and average intensity

maartenhammer committed:
#1225-complete overhaul of PDF Generator

maartenhammer committed:
#1225 fixed typo in PDF generator

maartenhammer committed:
#1225 Added an option to add scans to PDF

maartenhammer committed:
#1225 Added multipage

maartenhammer committed:
#1225 added blocks.

maartenhammer committed:
#1225 Removed SPM-Figure, Fixed Blocks"

maartenhammer committed:
#1225 added better font scaling in pdf generator

maartenhammer committed:
#1225 removed unneccesary files from GeneratePDF

maartenhammer committed:
#1225 moved patients.tsv evaluation to a callable subfunction in qc_GenerateReport

maartenhammer committed:
#1225 GenerateReport restructured PrintImage to prepare for relative pathing

maartenhammer committed:
#1225 GenerateReport added relative path support for subject and population folder

maartenhammer committed:
#1225 Added lineSpacing and automatic text resize in GeneratePDF

maartenhammer committed:
#1225 Config Report Template is now a proper example

maartenhammer committed:
#1225 removed redundant visibility boolean in ReportGenerator

maartenhammer committed:
#1225 xASL_generatePDF now no longer requires arbitrary element names

maartenhammer committed:
#1225 removed debug line from xASL_GeneratePDF

maartenhammer committed:
#1225 Generate PDF: Changed page.number field to page.pageIdentifier field

maartenhammer committed:
#1225 xASL_Generate_Report minor figure handling changes

maartenhammer committed:
#1225 xASL_qc_GenerateReport added printing of headers to images and scans

maartenhammer committed:
#1225 xASL_qc_GenerateReport vastly improved the Block management

maartenhammer committed:
#1225 xASL_qc_GenerateReport added jvm check

maartenhammer committed:
#1225 xASL_qc_GenerateReport fixed crash when slice string was absent or empty

maartenhammer committed:
#1225 xASL_qc_GenerateReport reorganized and consolidated code

maartenhammer committed:
#1225 xASL_qc_GeneratePDF small template tweaks

maartenhammer committed:
#1225 xASL_QC_GenerateReport added string replace to recognize <WILDCARD> in json strings

maartenhammer committed:
#1225 xASL_QC_GenerateReport added default subject

maartenhammer committed:
#1225 Publish PDFGenerator.md

maartenhammer committed:
#1225 Update PDFGenerator.md

maartenhammer committed:
#1225 renamed subfuction xASL_QC_WildCardReplace

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+%   PDF generator
+%
+% -----------------------------------------------------------------------------------------------------------------------------------------------------
+% DESCRIPTION: This function checks output struct x and places the values
```
Please complete the description (chat GPT/Github colab can help ;) the title of this function suggests that this function generates the config file that is needed to generate the PDF? But that is not what is described here?

jan-petr replied:
Yes, it says what is done, but not for what purpose.

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+% 
+% EXAMPLE: xASL_qc_GeneratePDFConfig(x);
+% __________________________________
+% Copyright (C) 2015-2021 ExploreASL
```
2023

maartenhammer replied:
Old File removed

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+
+
+if nargin <2 || isempty(jsonPath)
+    jsonPath = strcat(x.dir.xASLDerivatives, '/ConfigReportPDF.json');
```
better to use fullfile(XX, YY) to ensure that this also works in Windows, as fullfile chooses / or \ dependent on the OS

maartenhammer replied:
Old File removed

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+%  Admin
+
+
+if nargin <2 || isempty(jsonPath)
```
You first need to check the first input, if there is no X this function will crash. That is fine, but we can have a specific error message with something like "Did you forget to load ExploreASL?"

maartenhammer replied:
Old File removed

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+ 
+try
+    %% Load Standard Configuration
+    sPath = strcat(x.opts.MyPath, '/Functions/QualityControl/templateConfigReportPDF.json');
```
fullfile as above

maartenhammer replied:
Old File removed

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+    %% Load Standard Configuration
+    sPath = strcat(x.opts.MyPath, '/Functions/QualityControl/templateConfigReportPDF.json');
+    
+    if ~isfile(sPath)
```
isn't exist(sPath, 'file') better?

maartenhammer replied:
Old File removed

maartenhammer replied:
But yes, replaced on different locations!

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+    sPath = strcat(x.opts.MyPath, '/Functions/QualityControl/templateConfigReportPDF.json');
+    
+    if ~isfile(sPath)
+        warning('Default PDF configuration not found, check the explore ASL directory.')
```
ExploreASL, not explore ASL ;)

be specific: what does it mean to "check the ... directory"? Does that mean that the user should go to that directory and check its contents? Or that the correct ExploreASL directory is provided as input to this function? Or what does "check" mean?

maartenhammer replied:
Old File removed

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+    
+    if ~isfile(sPath)
+        warning('Default PDF configuration not found, check the explore ASL directory.')
+        return
```
I think we already catch errors in ExploreASL's xASL_Iteration etc, so you might want to simply use error instead of warning&return

maartenhammer replied:
Changed in other files

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+        config = xASL_adm_LoadPDFConfig(sPath);
+    end 
+
+    clear OutputFields 
```
I did this previously as well but Paul warned me that using "clear" in Matlab functions is not a nice way to do it, better to replace the variable with an empty one or just ignore it.
In any case, here "clear" is redundant because you redefine it on the next line

maartenhammer replied:
Old File removed

maartenhammer replied:
But good to know for future projects

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+% Copyright (C) 2015-2021 ExploreASL
+
+%% -----------------------------------------------------------------------------------------------
+%  Admin
```
Nice function! You're lacking comments here and there. Also, there is an admin section here, but no other sections. I always try to improve readability by using sections and comments. You could just add another section after the admin (checking input arguments, paths etc)

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+function [Config] = xASL_adm_LoadPDFConfig(x)
+% xASL_adm_LoadPDFConfig Loads parameters from a .json config file
+% Essentialy an alias of spm_jsonread.
```
Essentialy -> Essentially

jan-petr replied:
Is it just an alternative for jsonread? I guess it's much more than that. So this description should be improved. Also - lets not mention spm_jsonread, but just "reading JSON" or something like that.

maartenhammer replied:
Description improved, indeed not just an alias. It has some Qc/warnings/Error handling.

jan-petr replied:
OK

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+% FORMAT: [config] = xASL_adm_LoadPDFConfig(x)
+%
+% INPUT:
+%   ConfigPath  - path to json REQUIRES
```
REQUIRES -> REQUIRED

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+%   ConfigPath  - path to json REQUIRES
+%
+%   OUTPUT:
+%   config      - Struct containing all parameters of the jsonfile
```
REQUIRED or OPTIONAL?

jan-petr replied:
outputs are always optional ;) In a way, so no need to mention. Right?!

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+%
+% INPUT:
+%   x           - structure containing fields with all information required to run this submodule (REQUIRED)
+%   jsonPath    - Path where json Should be placed  (OPTIONAL DEFAULT is .../derivatives/ExploreASL/ConfigReportPDF.json)
```
comma missing between OPTIONAL, DEFAULT

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+% INPUT:
+%   x           - structure containing fields with all information required to run this submodule (REQUIRED)
+%   jsonPath    - Path where json Should be placed  (OPTIONAL DEFAULT is .../derivatives/ExploreASL/ConfigReportPDF.json)
+%   bPrintInvisible - boolean specifying wether you want the config file to contain invisible parameters as well
```
OPTIONAL or REQUIRED?

jan-petr replied:
OPTIONAL, DEFAULT = false.

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+% -----------------------------------------------------------------------------------------------------------------------------------------------------
+% EXAMPLE: [config] = xASL_adm_LoadPDFConfig(x);
+% __________________________________
+% Copyright 2015-2022 ExploreASL
```
2023

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+
+
+%% ------------------------------------------------------------------------
+% 0. Argument check
```
Where is section 1? Now it jumps from 0 to 2

jan-petr replied:
it seems that "load JSON" section is 2 below instead of 1.

maartenhammer replied:
Fixed descriptions

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+
+%% ------------------------------------------------------------------------
+% 0. Argument check
+Config = struct; % default
```
Nice, this is better than `clear Config` (which you used above)

maartenhammer replied:
This was a little redundant, so I removed it instead. 
Used to be a safeguard as to not return an empty config, which is no longer possible.

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+% 0. Argument check
+Config = struct; % default
+
+if nargin<1 || isempty(x)
```
Nice, this was missing above

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+    return
+end
+
+ConfigPath = strcat(x.dir.xASLDerivatives, '/ConfigReportPDF.json');
```
fullfile, as above

maartenhammer replied:
Replaced and fixed

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+
+%% ------------------------------------------------------------------------
+%% 2. Load JSON file 
+if isfile(ConfigPath)
```
same as above, not exist?

jan-petr replied:
it seems that "load JSON" section is 2 below instead of 1.

maartenhammer replied:
Fixed description and exists is faster and better!

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+if isfile(ConfigPath)
+    Config = spm_jsonread(ConfigPath);
+else
+    fprintf('configReportPDF.json does not exists, using default config \n');
```
I would print the full path that doesn't exist, not only the filename

maartenhammer replied:
Changed to be more Specific

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+    Config = spm_jsonread(ConfigPath);
+else
+    fprintf('configReportPDF.json does not exists, using default config \n');
+    standardPath = strcat(x.opts.MyPath, '/Functions/QualityControl/templateConfigReportPDF.json');
```
fullfile

maartenhammer replied:
Resolved and Fixed

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+%% ------------------------------------------------------------------------
+%% 3. Deal with warnings
+if isempty(fields(Config))
+    warning('config struct empty, Check if config JSON is configured correctly.');
```
Always good to be specific. If the user gets this warning on the screen, there is no clue as to which config file this is. So better to specify e.g. "check that the configuration JSON of the PDF report is configured correctly"

maartenhammer replied:
Changed to be more Specific

HenkMutsaerts commented on Functions/QualityControl/templateConfigReportPDF.json:
```diff
@@ -0,0 +1,338 @@
+                    "text": "This is an sample for the Explore ASL report Generator, See ExploreASL/Function/templateConfigReportPDF.json for the script generating this PDF."
+                },{
+                    "type": "scan",
+                    "name": "Pop_Path_rT1",
```
Nice! Only I would program the overlays with identical config layout as the "underlay", now the underlay has type (and overlay doesn't), underlay has the filename as value of the key "name" and overlay has the filename itself as key (with contour opacity color as children).

So I would rather say something like:
type: scan
position
size
slice
layer 1: {
name
contour
opacity
color}
layer 2: {

name
contour
opacity
color}
layer 3: {

name
contour
opacity
color}

etc. There is some redundancy in my approach, but having all layers identically set up as brother/sister layers makes programming easier I would say

HenkMutsaerts commented on Functions/QualityControl/templateConfigReportPDF.json:
```diff
@@ -0,0 +1,338 @@
+                        }
+                    }
+                },{
+                    "type": "patients"
```
what type is patients?

HenkMutsaerts commented on Functions/QualityControl/templateConfigReportPDF.json:
```diff
@@ -0,0 +1,338 @@
+                    "type": "text",
+                    "settings": {
+                        "fontSize": "10",
+                        "fontWeight": "bold"
```
So, I guess these settings only apply to this text type, not the other ones?
And the other ones use the default settings?

maartenhammer replied:
Yes, if settings are specified for something they are only for that specific field. If you want to swap the defaults you should add a "type":"settings" json field.

maartenhammer replied:
Change Settings -> TextSettings for clarity

HenkMutsaerts commented on Functions/QualityControl/templateConfigReportPDF.json:
```diff
@@ -0,0 +1,338 @@
+                    "type": "text",
+                    "text": "If the QC value is outside your specified range, it will display in red."
+                },{
+                    "type": "qc",
```
We need to define "type" and it is best to have a sensible classes/groups/levels

E.g., "text" and "image" (not scan) are datatypes
but "qc" and "patient" are not datatypes, but what are they?

maartenhammer replied:
Change to contentImage2D, contentImage3D, contentValues, contentPatients, metaPage, metaTextSettings, metaBox/metaBlock

HenkMutsaerts commented on Functions/QualityControl/templateConfigReportPDF.json:
```diff
@@ -0,0 +1,338 @@
+                            "text": "This block has text on the left and scans on the right."
+                        },{
+                            "type": "image",
+                            "popPath": "M0Check/Cor_M0_<SUBJECT>_<SESSION>.jpg",
```
Note that SESSION becomes RUN with BIDS (and visit becomes SESSION)

maartenhammer replied:
x.SUBJECT and x.SESSION (should be x.RUN in BIDS) and x.visit (should be x.SESSION in BIDS)

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+function [x,config] = xASL_qc_GenerateReport(x, subject)
```
Why do you output config?

maartenhammer replied:
Debugging only, removed.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+
+    if ispc
+        settings.fontSize = 10;
+    elseif isunix || ismac
```
mac == unix, so you can remove ismac

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+        settings.fontSize = 10;
+    elseif isunix || ismac
+        settings.fontSize = 8;
+    end 
```
else: "unknown OS, warning" is missing

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+
+%% ================================================================================================================================================
+%% ================================================================================================================================================
+function [settings] = xASL_sub_parseJson(json, x, figure, line, settings)
```
This is quite a long file so I would create a separate function and wrapper (xASL_qc_GenerateReport is the wrapper, and changes minimally with your PR right? That makes it nicely clear)

Also, I thought we decided to call this e.g., xASL_qc_GenerateReport_sub_parseJson because otherwise it is not clear where the subfunction is

maartenhammer replied:
Moved all subfunctions to xASL_qc_ParsePdfConfig (maybe xASL_adm_ instead?) and renamed the subfunctions.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+fprintf('Printing ExploreASL PDF report:   \n');
+
+% Parse the entire Json Stack automatically making all the pages.
+[~] = xASL_sub_parseJson(config, x, [], [], settings);
```
You can remove the ~ output with Matlab

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+end
+
+% Set defaults
+settings = xASL_sub_defaultSettings();
```
name of the subfunction, see below

maartenhammer replied:
Changed, but this does make the function names very long.  Will resolve all other occurances of this issue to keep conversation condensed.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+
+%% ================================================================================================================================================
+%% ================================================================================================================================================
+function [settings] = xASL_sub_parseJson(json, x, figure, line, settings)
```
json, figure, line are not defined in the main function.
Also, some may be protected Matlab variables, like figure, that are better to give a unique name

maartenhammer replied:
Added description of the JsonParser, including defining undefined variables.

Changed figure to currentFigure, settings to settingsPDF.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+function [settings, line] = xASL_sub_parseContent(currentField, x, figure, line, settings)
+
+    if ~isstruct(currentField)
+        return;
```
And then? A warning?

maartenhammer replied:
This is just to ignore any nonStruct objects passed to this function this can happen in normal operation because the every top-level Json-field is at some point passed to this. I could change the caller to be more strict if wanted.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    end
+
+    if ~isfield(currentField, 'type')
+        error([currentField, 'containts no field specifying the type of content']);
```
containts -> contains

This error lacks a hint to the user what could have gone wrong

maartenhammer replied:
fixed and added better hint

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+            xASL_sub_printPage(currentField, x, settings);  
+        case 'block'
+            xASL_sub_printBlock(currentField, x, figure, settings);
+        case 'settings'
```
shouldn't it be printSettings? for consistency

I like this function!

maartenhammer replied:
Settings arent printed to the page, which is why theres a difference in name.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+
+function  xASL_sub_printPage(json, x, settings)
+
+    clear primaryFig ax  
```
see above, better not clear but define empty defaults

maartenhammer replied:
removed clear, replaced in certain instances with cla before defining images to have a clear figure.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    primaryFig = figure('visible', 'off','Units', 'centimeters','Position', [0 0 21 29.7]);
+
+    %% Print the Title
+    logoPath = fullfile(x.opts.MyPath, 'Design/ExploreASL_logoHeader.png');
```
/ will crash on Windows, better to replace this by

logoPath = fullfile(x.opts.MyPath, 'Design', 'ExploreASL_logoHeader.png');

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+
+    %% Print the Title
+    logoPath = fullfile(x.opts.MyPath, 'Design/ExploreASL_logoHeader.png');
+    [~] = xASL_sub_PrintImage(logoPath, [], primaryFig, settings, [0 0.96 1 0]);
```
~ not needed

by the way if we use variables like
figPrimary
pathLogo

it is easier to list them alphabetically than having 

primaryFig
logoPath

(if we have multiple figs paths etc)

maartenhammer replied:
I am not 100% sure what is meant here.
Do you mean change the variables to <type><Identifier> (figPrimary instead of primaryFig) or do you mean in order or definitions?
For both of these variables only one exists at a time.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    %% Print the Footer
+    xASL_sub_PrintText('This report was automatically generated by ExploreASL', primaryFig, [0 0.02 1 0], settings);
+
+    %% Parse JSON and create the file
```
Parse which JSON and create which file? Be specific

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    %% Parse JSON and create the file
+    [~] = xASL_sub_parseJson(json, x, primaryFig, [0 0.93 1 0], settings);
+
+    PrintFile = ['xASL_Report_', json.pageIdentifier];
```
Do we print each page as separate file?

maartenhammer replied:
Yes, limitation of matlab printing.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    clear primaryFig ax
+end
+
+function xASL_sub_printBlock(input, x, pageFig, settings)
```
Also the subfunctions should have headers (at least to make them a bit readable)

and all functions are missing a lot of comments to make them readable

maartenhammer replied:
Added!

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+function xASL_sub_printBlock(input, x, pageFig, settings)
+    position = xASL_str2num(input.position);
+    size = xASL_str2num(input.size);
+    [settings.canvas, line] = xASL_sub_createNewCanvas(position, size, settings.canvas);
```
why is canvas in settings but position size line etc are not?

maartenhammer replied:
Position and size are content specific parameters to draw new canvasses, added more context to it.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+        return
+    end
+
+    sParticipants = xASL_tsvRead(ExistsParticipants{1});
```
what does s mean?
n usually means number

maartenhammer replied:
XASL_tsvRead generates a struct from the tsv changed it to structParticipants

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    end
+
+    % Use field specific settings if they exists
+    settings.fontName = 'monospace';
```
I thought all settings were in the default config file, but some are in this code? Should we be consistent?

maartenhammer replied:
Yes, but qc-values specifically I prefer to use a different font to the standard default. 
Monospace ensures that the numbers line up better regardless of what comes before it. 
I have explained it better in the comments.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+        json.range = '';
+    end
+
+    if isnumeric(TempValue)
```
You can use xASL_num2str which deals with isnumeric and different formats of numbers

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+        if isfield(input, 'header')
+            header = input.header;
+        else
+            header = '';
```
Isn't it easier to default header to '' above, that avoids repetition

maartenhammer replied:
Yes

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+        header = '';
+    end
+
+    [canvas, ~] = xASL_sub_createNewCanvas(position(1:2), position(3:4), settings.canvas);
```
~ can be removed. is only necessary when wanting to take the second or other output but ignoring the first

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    end
+
+    [canvas, ~] = xASL_sub_createNewCanvas(position(1:2), position(3:4), settings.canvas);
+    ax=axes('Position', canvas, 'Visible', settings.axesVisible, 'Parent', figure);
```
space missing with =

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+
+    [canvas, ~] = xASL_sub_createNewCanvas(position(1:2), position(3:4), settings.canvas);
+    ax=axes('Position', canvas, 'Visible', settings.axesVisible, 'Parent', figure);
+    [img, ~, alphachannel] = imread(ImagePath);
```
Careful! imread and imshow are from the image processing toolbox. All ExploreASL functions should work without Matlab toolboxes

jan-petr replied:
Yes please. Cause not all users have them and hopefully, we will now be able to compile without toolboxed.

maartenhammer replied:
I decided to check this, (discussing with Jan in Whatsapp) I have checked many times, but Imshow and Imread don't trigger any toolbox licenses. with `licence('inuse')` I regularly check if any toolboxes are triggered and this function does not trigger any. Using Matlabs built in dependency report you can also see no external or toolbox dependencies. 
![image](https://github.com/ExploreASL/ExploreASL/assets/40172631/440b95c2-c0b5-43e0-b270-1671afb346af)

jan-petr replied:
I confirm that - imread is not in Image processing toolbox anymore. Imshow is in the toolbox (or at least stated on Matlab webpages), but  mysteriously, this doesn't trigger the license demand. 

Both functions are in toolbox/matlab/images and not in toolbox/images

So it is all fine. We can use imshow and imread without restrictions!

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    [canvas, ~] = xASL_sub_createNewCanvas(position(1:2), position(3:4), settings.canvas);
+    ax=axes('Position', canvas, 'Visible', settings.axesVisible, 'Parent', figure);
+    [img, ~, alphachannel] = imread(ImagePath);
+    fg= imshow(img);
```
Best to avoid abbreviations, especially unreadable ones ;) Your code should be readable to others

maartenhammer replied:
This isnt an abreviation but a matlab reserved function I believe, I don't ever use the ax/fg variables.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+
+    ImIn = {x.P.(input.name)};
+    header = input.name;
+    ax=axes('Position', canvas, 'Visible', settings.axesVisible, 'Parent', figure);
```
space =

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    header = input.name;
+    ax=axes('Position', canvas, 'Visible', settings.axesVisible, 'Parent', figure);
+    
+    if isfield(input, 'overlay')
```
I don't understand, a slice is a subset of a full image, and thus of an overlay. Does that mean that this function has multiple functions depending on its input? Because that would be confusing... Unless you have a single function that handles all inputs, but now it only handles overlay and slice inputs, we should avoid functions becoming convoluted.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+        x.S.CorSlices = [25, 50, 90];
+    end
+
+    img = xASL_vis_CreateVisualFig(x, ImIn, [], [], [], [], [], [], [], [], [], [], []);
```
avoid abbreviations

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    end
+end
+
+function [newCanvas, line] = xASL_sub_createNewCanvas(position, size, oldCanvas)
```
should oldCanvas not be the first input argument, similar to newCanvas being the first output argument?

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+        slice = [];
+    end
+end
+
```
I would add some extra lines with %% ==================== between the sub functions to separate them clearer/more readable (at least I like to make functions readable :)

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+end
+
+function [newCanvas, line] = xASL_sub_createNewCanvas(position, size, oldCanvas)
+    newPosition = [oldCanvas(1) + position(1) * oldCanvas(3) oldCanvas(2) + position(2) * oldCanvas(4)];
```
There are comments lacking, e.g., what is a "canvas" in your code?

maartenhammer replied:
Added

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+            error('xASL_sub_PrintText couldnt find string of printable text')
+    end
+
+    ax=axes('Position', line ,'Visible', settings.axesVisible, 'Parent', figure);
```
check spaces ax=axes

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    line = [newPosition(1)  newPosition(2) + newSize(2) newSize(1) 0];
+end
+
+function line = xASL_sub_PrintText(input, figure, line, settings)
```
above you have clearly defined types, and now you have an undefined first argument "input", which is handled differently depending on its content, I would avoid this, this makes for difficult to read code

maartenhammer replied:
This is mostly to avoid duplication, as of right now the watermark (This is generated by ExploreASL on the bottom of the page) is handled by the same function as the function that parses the pdf report. We can discuss this later.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    line(2) = line(2) - (settings.lineSpacing) - (settings.fontSize * 0.001);
+
+    if line(2) < 0 
+        warning('No space left on page!');
```
On which page? Make warnings clear

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    if line(2) < 0 
+        warning('No space left on page!');
+    elseif line(2) < settings.canvas(2) 
+        warning('Printing outside canvas, check block settings.');
```
Which canvas? Which block settings? Help the user

maartenhammer replied:
As of right now this is information not know to the program in this subfunction, same of comment above.

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+
+    position(4) = 0;
+    settings.HorizontalAlignment = 'center';
+    figurecount = settings.figurecount + 1;
```
why is figureCount (check camelCase) defined twice, what is the role of "settings"? Make it readable

HenkMutsaerts reviewed (changes requested):
Very nice work! Some minor stuff, that are mainly about code, header, and comment consistencies, trying to keep everything readable. You currently have a lot of subfunctions where variables change names (e.g., "input") and depending on the content of the variable a different action will be triggered by a single function, that may not help readability and debugging. I would always go for simplicity, like you nicely do sometimes with things like

switch type
               A -> do this subfunction
               B -> do this subfunction

instead of doing the same subfunction for both A and B but inside the subfunction determining to what other subfunction A or B go

Another important issue: you use imshow and perhaps other functions from the Matlab image processing toolbox. Jan and I have put a lot of effort in not having any toolbox dependencies outside Matlab itself. This will ensure that most users can use ExploreASL (also those that didn't pay for the toolbox) and that the docker container stays lean.

jan-petr commented on Functions/Administration/xASL_adm_GeneratePDFConfig.m:
```diff
@@ -0,0 +1,83 @@
+        end
+    end
+
+    spm_jsonwrite(jsonPath, OutputStruct, struct('indent', ' '));
```
Please use xASL_io_WriteJson or xASL_io_ReadJson
the SPM version had several flaws and was also quite lenient in checking if JSONs were formally correct.
So that's why we wrote our wrapper function around jsonencode functions of Matlab.

maartenhammer replied:
Fixed

jan-petr commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+%% ------------------------------------------------------------------------
+%% 2. Load JSON file 
+if isfile(ConfigPath)
+    Config = spm_jsonread(ConfigPath);
```
replace with xASL_io_ReadJson

maartenhammer replied:
Replaced and resolved, works exellently.

jan-petr commented on Functions/Administration/xASL_adm_LoadPDFConfig.m:
```diff
@@ -0,0 +1,54 @@
+else
+    fprintf('configReportPDF.json does not exists, using default config \n');
+    standardPath = strcat(x.opts.MyPath, '/Functions/QualityControl/templateConfigReportPDF.json');
+    Config = spm_jsonread(standardPath);
```
xASL_io_ReadJson

maartenhammer replied:
same as Above

jan-petr commented on Functions/QualityControl/templateConfigReportPDF.json:
```diff
@@ -0,0 +1,338 @@
+{
```
please note that xASL_io_ReadJson strictly adheres to JSON standard - unlike spm_jsonread.

So don't be surprised if reading some JSONs throws an error.

maartenhammer replied:
It didn't!

jan-petr commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+function [x,config] = xASL_qc_GenerateReport(x, subject)
+% xASL_qc_GenerateReport Print output QC
+%
+% FORMAT: xASL_qc_GenerateReport(x)
```
[, subject]) is missing here

jan-petr replied:
also - is there is is there not output?

HenkMutsaerts replied:
Why did you resolve this conversation? Did you understand his comment (it is about line 4, not line 1)?

jan-petr commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+%   x           - structure containing fields with all information required to run this submodule (REQUIRED)
+%   subject     - subject name (OPTIONAL, default = x.SUBJECT)
+%
+% OUTPUT: n/a
```
function definition on the first line contains outputs.

maartenhammer replied:
Added

jan-petr commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    config = NaN;
+    return;
+end
+
```
please add check for the first input parameter

maartenhammer replied:
Resolved and Added

jan-petr commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,418 @@
+    end
+
+    img = xASL_vis_CreateVisualFig(x, ImIn, [], [], [], [], [], [], [], [], [], [], []);
+    fg = imshow(img);
```
again imshow

jan-petr reviewed (changes requested):
Some minor changes needed.

maartenhammer committed:
#1225 xASL_adm_LoadPDFConfig cleanup, documentation and replaced functions to native ExploreASL functions

maartenhammer committed:
#1225 xASL_qc_GenerateReport fixed spacing after punctuation

maartenhammer committed:
#1225 renamed subfunction from xASL_sub_ to xASL_qc_GenerateReport_sub_

maartenhammer committed:
#1225 Changed xASL_qc_GenerateReport to wrapper, renamed subfunctions and variables to xASL_qc_ParsePdfConfig

maartenhammer committed:
#1225 xASL_qc_ParsePdfConfig renamed json/input to <type>Struct to improve specifity of subfunctions

maartenhammer committed:
#1225 xASL_qc_ParsePdfConfig added more comments

maartenhammer committed:
#1225 xASL_qc_Parse_PdfConfig added comments to subfunctions.

maartenhammer committed:
#1225 xASL_qc_parsePdfConfig comment changes

maartenhammer commented on Functions/QualityControl/templateConfigReportPDF.json:
```diff
@@ -0,0 +1,338 @@
+                        "Pop_Path_SliceGradient": {
+                            "contour": "true",
+                            "opacity": "0.3",
+                            "color": "r"
```
```
"type": "contentImage3D",
                    "name": "Pop_Path_rT1",
                    "position": "[0.70 0.35]",
                    "size": "[0.3 0.3]",
                    "slice": {
                        "TraSlice": "[25 75]",
                        "CorSlice": "[25 75]",
                        "SagSlice": "[]"
                    },
                    "overlay": [{
                            "name":"Pop_Path_SliceGradient",
                            "contour": "true",
                            "opacity": "0.3",
                            "color": "r"
                            },{
                            "name":"Pop_Path_<layer2>",
                            "contour": "true",
                            "opacity": "0.3",
                            "color": "r"
                            }]

```

HenkMutsaerts committed:
#1225 xASL_qc_GenerateReport: bugfix x.SUBJECT

HenkMutsaerts committed:
#1225 xASL_qc_GenerateReport: bugfix x.SESSION

HenkMutsaerts commented:
Just tested this:

` [x, config] = xASL_qc_GenerateReport(x[, subject])` as you state above is not working, there is only a single x output.

Then I saw 2 minor bugs which I fixed.

Finally, the current config gives an example, can you replace the default config with one that creates a similar PDF as we had? Then we can improve it from there, but then at least I can already check how it works.

Also, now if there are any pre-existing xASL reports, it will delete them; this behavior should be disabled. Only, if there already is a PDF existing with an identical filename, it should perhaps rename the new file (I wouldn't touch the old/pre-existing file).

Also, this function now is dependent on `x.P`, so if a filename exists but is not defined in `x.P`, it will not be printed. I would instead just use relative filenames, not `x.P` fields (those fields are only useful within ExploreASL).

HenkMutsaerts committed:
#1225 xASL_qc_GenerateReport: bugfix x.P

HenkMutsaerts committed:
#1225 xASL_qc_GenerateReport: trying tracking

HenkMutsaerts committed:
#1225 xASL_qc_GenerateReport: disable figure in cli

maartenhammer commented:
> Just tested this:
> 
> ` [x, config] = xASL_qc_GenerateReport(x[, subject])` as you state above is not working, there is only a single x output.
> 
> Then I saw 2 minor bugs which I fixed.
> 
> Finally, the current config gives an example, can you replace the default config with one that creates a similar PDF as we had? Then we can improve it from there, but then at least I can already check how it works.
> 
> Also, now if there are any pre-existing xASL reports, it will delete them; this behavior should be disabled. Only, if there already is a PDF existing with an identical filename, it should perhaps rename the new file (I wouldn't touch the old/pre-existing file).
> 
> Also, this function now is dependent on `x.P`, so if a filename exists but is not defined in `x.P`, it will not be printed. I would instead just use relative filenames, not `x.P` fields (those fields are only useful within ExploreASL).

Excellent, the changes added by you seem to improve stability when parameters are missing, as well as some householding data. 

I will also improve this by including [the definitions here](https://ars.els-cdn.com/content/image/1-s2.0-S1053811920305176-mmc1.pdf) as default for units and names if no alternatives are given.

jan-petr committed:
#1225 xASL_qc_ParsePdfConfig: Header fix

jan-petr reviewed (approved):
Good work. No further comments.

maartenhammer committed:
#1225 GeneratePDF_Report added translation function BIDS->ExploreASL Legacy

maartenhammer committed:
#1225 Generate PDF report changed 'settings' to 'textSettings' for clarification

maartenhammer committed:
#1225 xASL_qc_GenerateReport added category/type distinction

maartenhammer committed:
#1225 GeneratePDF added qc default translations and rudimentory implementation

maartenhammer committed:
#1225 Changed default PDF report to look the same as old PDF Report (Text)

maartenhammer committed:
#1225 Changed default PDF report to look the same as old PDF Report (Images)

maartenhammer commented:
Last conversations have been handled:
Translation layer added that translates ExploreASL Legacy to BIDS when turned on
Settings has been renamed to TextSettings for clarity
Image has been renamed to Image2D 
scan has been renamed to Image3D
A secondary categorizer has been added (instead of a json item being Type = image2D, we now have category = "content" , type = "image".
This is to distinguish metadata json-objects (pages, blocks, settings) and content json objects(like image2d, image3d, text, QCValues).
Added translation layer to always automatically translate [some of the qc parameters here](https://ars.els-cdn.com/content/image/1-s2.0-S1053811920305176-mmc1.pdf). Not all have been filled in, discuss with either Jan/Henk Jan which should be filled in. 
Changed default print to look more like the example given in [figure 3 of the ExploreASL paper](https://www.sciencedirect.com/science/article/pii/S1053811920305176) .
[New default now looks like this](https://github.com/ExploreASL/ExploreASL/files/13441564/xASL_Report_Example_1.pdf)

Edit: 
Discussion with Henk-Jan on thursday.
Possible additions:
Pdfconfig json generator that reads x.output and populates a PDFconfig.json instead of using a default template in Functions->Quality Controll, add it to wrapper.
Add text elipses (...) when qc_param exceeds certain length. (added)

maartenhammer committed:
#1225 xASL_generateReport added string width limiting and ellipsis

maartenhammer committed:
#1225 xASL_num2str fixed bug prevention translation of logical variables

maartenhammer commented:
While adding text wrapping I've noticed that xASL_num2str does not convert logical to str, which the matlab native num2str does.
A minor change to xASL_num2str is added in [3e1290d](https://github.com/ExploreASL/ExploreASL/pull/1475/commits/3e1290d25693845dacfbfd9814bbbd3439430f2d) which fixes this, as I assume this is a bug instead of intended behaviour.  (Discussed with Jan over whatsapp).

jan-petr reviewed (changes requested):
I am merely blocking the merging, because some commit messages need fixing ;)

jan-petr committed:
#1225 Move PDFGenerator.md to docu

HenkMutsaerts commented on Functions/QualityControl/qc_translation.tsv:
```diff
@@ -0,0 +1,112 @@
+parameter	Alias	Unit	Range
```
As we now discussed, please add a qc_translation.txt with a brief explanation of what this file does, and do a spelling check. I can complete the aliases etc.

maartenhammer replied:
For these translation tables I'd rather keep it inside the exploreASL directory, as these wouldn't ever change. The only difference is which ones you use. Generating this in a subject/study directory would just generate a bunch of useless Tsv files for the user, and in the JSON you can override any and all of these anyway. 

Looking up BIDS definitions I do agree on.

HenkMutsaerts replied:
? Did you do/check this?

maartenhammer replied:
Yes, Bids does not have shortnames other than the standard bids names for parameters, I've added certain BIDS specific terms and renamed this to qc_glossary as a more "clear" name as for what this is. I've also spell checked and added a new row of "Descriptions", which can be used for longer descriptions, I've shortened the aliasses to make printing less ugly. 
Added a .txt file as well.

HenkMutsaerts commented on Functions/QualityControl/templateConfigReportPDF.json:
```diff
@@ -0,0 +1,695 @@
+{
+    "globalSettings": {
+        "category": "metadata",
```
Can we create this dynamically by default, based on the QC values in x.output? To be overridden if the user provides their own version?

maartenhammer replied:
We can now.

HenkMutsaerts replied:
?

HenkMutsaerts replied:
Maarten made this dynamic already, so this is only an example json.
@maartenhammer then please rename this to append the suffix "example" to this file name, so this is clear.

maartenhammer replied:
As discussed, this is now possible using the xASL_adm_GeneratePDFConfig function, which is called and the default behaviour of GenerateReport now. 
It essentially dumps all x.Output into a json configuration with minimal formatting, each series is it's own page due to matlab limitations. (Eg, an ASL pdf and T1 pdf are generated.)

HenkMutsaerts reviewed (changes requested):
Nice work! Some questions as we discussed, to keep the current functionality of generating the PDF dependent on what was analyzed in ExploreASL.

E.g, if you have analyzed 2 ASL runs, you want to show 2 ASL runs. If you have analyzed only T1w and FLAIR, you want to only show them.

So while there might be a static part of the template/settings file, it would be nice if these things can also be generated dependently.

maartenhammer committed:
#1225 Added script that automatically generates a PDF report config (qc_values)

maartenhammer committed:
#1225 Added script that automatically generates a PDF report config (qc_values)

maartenhammer committed:
#1225 Added script that automatically generates a PDF report config (qc_images)

HenkMutsaerts commented on Functions/Administration/xASL_adm_LoadPdfConfig.m:
```diff
@@ -0,0 +1,58 @@
+
+%% ------------------------------------------------------------------------
+%% 2. Load JSON file 
+if (exist(configPath, 'file') == 2)
```
() brackets are not required, and you don't have to test if the output of exist equals 2, this is what 'file' does

maartenhammer replied:
Fixed

HenkMutsaerts commented on Functions/Administration/xASL_adm_GeneratePdfConfig.m:
```diff
@@ -0,0 +1,140 @@
+function [config] = xASL_adm_GeneratePdfConfig(x, subject, bOverwrite)
```
Would rename all QC-related functions to xASL_qc_ and move to the QualityControl folder, just easier to have these all sorted.

maartenhammer replied:
Okay I will do that

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,122 @@
+%   xASL_Report_SubjectName.pdf - printed PDF rapport containing QC images & values
+%
+% -----------------------------------------------------------------------------------------------------------------------------------------------------
+% DESCRIPTION:  This function generates a PDF report based on a predifined Json configuration file.
```
predefined

maartenhammer replied:
fixed

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_GenerateReport.m:
```diff
@@ -0,0 +1,122 @@
+function [x, settings] = xASL_qc_GenerateReport(x, subject)
+% xASL_qc_GenerateReport Generates a PDF report based on a predifined Json configuration file
```
predefined

maartenhammer replied:
fixed

HenkMutsaerts commented on Functions/QualityControl/xASL_qc_ParsePdfConfig.m:
```diff
@@ -0,0 +1,675 @@
+    
+    % Pad spaces
+    if ~isempty(textToPrint)
+        textToPrint = [textToPrint ' '];
```
Here we sometimes have char arrays that are not only horizontal, because a parameter can have multiple values apparently, and then the values should be concatenated in a string/char or just the first reported (depending on the length).

E.g., I had here:

textToPrint =

  2×4 char array

    '1710'
    '3142'

First, I would never use char arrays, can we get rid of them?

Because if we have 

textToPrint2 = {'1710';'3142'}

textToPrint2 =

  2×1 cell array

    {'1710'}
    {'3142'}

this is easy to concatenate iterating over cells (2 elements)
With a char array, this example would be 8 elements.

Char arrays is something new of Matlab I think, at least Jan and I never used them.

Or if it becomes very tricky to get rid of char arrays, perhaps you can think of a different solution?

maartenhammer replied:
I have decided to leave it as Char as it is, I've reworked the way it reads values, and it now horizontally concatenates them. Sadly for chars a simple var= var(:)' does not work due to indexing of chars, so I used append instead. This can become slow if a **very large** amount of values need to be concatenated.  This can be relieved by pre-defining the character array length and replacing values inside a predefined array, but realistically it won't ever matter for this purpose.

HenkMutsaerts reviewed (changes requested):
Some minor comments, and a bug while testing ExploreASL's TestDataSet

HenkMutsaerts commented on Functions/QualityControl/templateConfigReportPDFAlternative.json:
```diff
@@ -0,0 +1,457 @@
+{
```
Same for all files, if they are now dynamically created, please add the suffix "example" or something so this is clear.

maartenhammer replied:
Changed this to be "example"

HenkMutsaerts reviewed (changes requested):
Few additional things.

maartenhammer committed:
#1225 Generate PDF report moving functions, addition of comments, cleanup of code

maartenhammer commented:
Two force pushed to incorporate two minor appends to the last commit.

maartenhammer committed:
#1225 Generate Report Fixed bug when quality paramteres contain multiple values

HenkMutsaerts reviewed (approved):
My test worked fine. I have added some wishes to the follow-up issue #1476 (in which we actually implement this in ExploreASL).

jan-petr reviewed (approved):
OK
